app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt

```diff
@@ -0,0 +1,140 @@
+ * maintain consistency in the user experience.
+ *
+ * @property profilesViewModel Handles profile data and filtering
+ * @property meetingRequestViewModel Manages sending notifications
+ * @property appDataStore Manages user preferences and settings
+ * @property context Android context for system services
+ * @property filterViewModel Manages user's filter settings (radius, age, gender)
+ */
+class EngagementNotificationManager(
+    private val profilesViewModel: ProfilesViewModel,
+    private val meetingRequestViewModel: MeetingRequestViewModel,
+    private val appDataStore: AppDataStore,
+    private val context: Context,
+    private val filterViewModel: FilterViewModel
+) {
+  private var notificationJob: Job? = null
+  private val scope = CoroutineScope(Dispatchers.Main)
+
+  /** Start monitoring for nearby users with common tags */
+  fun startMonitoring() {
+    stopMonitoring() // Stop any existing monitoring
+
+    notificationJob =
+        scope.launch {
+          while (true) {
+            checkNearbyUsersForCommonTags()
+            delay(CHECK_INTERVAL)
+          }
+        }
+  }
+
+  /** Stop monitoring for nearby users */
+  fun stopMonitoring() {
+    notificationJob?.cancel()
+    notificationJob = null
+  }
+
+  private fun isAppInForeground(): Boolean {
+    val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
+    val appProcesses = activityManager.runningAppProcesses ?: return false
+    val packageName = context.packageName
+
+    for (appProcess in appProcesses) {
+      if (appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND &&
+          appProcess.processName == packageName) {
+        return true
+      }
+    }
+    return false
+  }
+
+  private fun checkNearbyUsersForCommonTags() {
+    val selfProfile = profilesViewModel.selfProfile.value ?: return
+    val selfLocation = selfProfile.location ?: return
+
+    // Use the user's selected radius from FilterViewModel
+    profilesViewModel.getFilteredProfilesInRadius(
+        center = selfLocation,
+        radiusInMeters = filterViewModel.selectedRadius.value,
+        genders = filterViewModel.selectedGenders.value,
+        ageRange = filterViewModel.ageRange.value)
+
+    // Launch a coroutine to collect the filtered profiles
+    scope.launch {
+      // Only proceed if we are discoverable
+      if (!appDataStore.isDiscoverable.first()) return@launch
+
+      profilesViewModel.filteredProfiles.collectLatest { nearbyProfiles ->
+        // Only send notifications if the app is not in foreground
+        if (!isAppInForeground()) {
+          processNearbyProfiles(selfProfile, nearbyProfiles)
+        }
+      }
+    }
+  }
+
+  private fun processNearbyProfiles(selfProfile: Profile, nearbyProfiles: List<Profile>) {
+    val selfTags = selfProfile.tags
+
+    for (nearbyProfile in nearbyProfiles) {
+      if (nearbyProfile.uid == selfProfile.uid) continue // Skip self
```

Comment:
Wouldn't here be a more elegant solution ? like using the Filter function of nerbyProfiles before calling the for loop ? Would make the code more readable and optimize a little bit the code. You could do something like "val newList = nearbyProfiles.filter{it != selfProfile}